File: record_android/android/src/main/java/com/llfbandit/record/record/stream/RecorderRecordStreamHandler.java
package com.llfbandit.record.record.stream;

import android.app.Activity;

import androidx.annotation.Nullable;

import io.flutter.plugin.common.EventChannel;

public class RecorderRecordStreamHandler implements EventChannel.StreamHandler {
  // Event producer
  private EventChannel.EventSink eventSink;

  @Nullable
  private Activity activity;

  @Override
  public void onListen(Object o, EventChannel.EventSink eventSink) {
    this.eventSink = eventSink;
  }

  @Override
  public void onCancel(Object o) {
    eventSink = null;
  }

  public void sendRecordChunkEvent(byte[] buffer) {
    if (eventSink != null && activity != null) {
      activity.runOnUiThread(() -> eventSink.success(buffer));
    }
  }

  public void setActivity(@Nullable Activity activity) {
    this.activity = activity;
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/record/stream/RecorderRecordStreamHandler.kt
package com.llfbandit.record.record.stream

import com.llfbandit.record.record.util.MainThread
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink

class RecorderRecordStreamHandler : EventChannel.StreamHandler {
  // Event producer
  private var eventSink: EventSink? = null

  override fun onListen(arguments: Any?, events: EventSink?) {
    this.eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  fun sendRecordChunkEvent(buffer: ByteArray) {
    MainThread.post {
      eventSink?.success(buffer)
    }
  }

  fun sendErrorEvent(ex: Exception) {
    MainThread.post {
      // `details` must be codec-encodable; a Throwable is not.
      eventSink?.error("-1", ex.message ?: ex.toString(), ex.cause?.toString())
    }
  }
}
